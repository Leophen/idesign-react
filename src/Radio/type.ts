export interface RadioGroupProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 单选框组选中的值
   */
  currentValue?: string | number;
  /**
   * 单选框组全局类型
   * @default radio
   */
  type?: 'radio' | 'radio-button';
  /**
   * 按钮单选框全局尺寸
   * @default medium
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * 单选框组是否全局禁用
   */
  disabled?: boolean;
  /**
   * 选中某一项时触发
   */
  onChange?: (value: string | number, context: React.ChangeEvent<HTMLInputElement>) => void;
}

export interface RadioProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 单选框类型
   * @default radio
   */
  type?: 'radio' | 'radio-button';
  /**
   * 按钮单选框尺寸
   * @default medium
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * 是否选中
   * @default false
   */
  checked?: boolean;
  /**
   * 是否禁用组件
   * @default false
   */
  disabled?: boolean;
  /**
   * 单选框的值
   */
  value?: string | number;
  /**
   * 值变化时触发
   */
  onChange?: (checked: boolean, context: React.ChangeEvent<HTMLInputElement>) => void;
}

export interface RadioContextValue {
  inject: (props: RadioProps) => RadioProps;
}
