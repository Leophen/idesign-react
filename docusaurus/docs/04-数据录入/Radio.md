---
title: Radio 单选框
---

多用于在一组选项中进行单项选择且选项互斥，一般需要提交的选择型控件。

## 引入

```js
import {Radio} from 'idesign';
```

## 基本用法

最简单的单选框形式。

```jsx live fff
function App() {
    const [checked, setChecked] = useState(false);

    const handleChange = val => {
        console.log(val);
        setChecked(val);
    };

    return (
        <div className="demo-column">
            <h5>无默认值</h5>
            <Radio>无默认值</Radio>

            <h5>有默认值（非受控）</h5>
            <Radio defaultValue={checked}>有默认值（非受控）</Radio>

            <h5>有固定值（受控）</h5>
            <Radio checked={checked}>有固定值（受控）</Radio>

            <h5>通用方法</h5>
            <Radio checked={checked} onChange={handleChange}>
                通用方法
            </Radio>
        </div>
    );
}
```

## 单选框组

使用 `Radio.Group` 可以包裹 `Radio` 形成一组单选框组。

```jsx live fff
function App() {
    const [selected, setSelected] = useState('2')

    const handleChange = (e) => {
        console.log(e)
        setSelected(e.target.value)
    }

    return (
        <div className="demo-column">
            <h5>无默认值</h5>
            <Radio.Group>
                <Radio value="1">选项一</Radio>
                <Radio value="2">选项二</Radio>
                <Radio value="3">选项三</Radio>
                <Radio value="4">选项四</Radio>
            </Radio.Group>

            <h5>有默认值（非受控）</h5>
            <Radio.Group defaultValue={selected}>
                <Radio value="1">选项一</Radio>
                <Radio value="2">选项二</Radio>
                <Radio value="3">选项三</Radio>
                <Radio value="4">选项四</Radio>
            </Radio.Group>

            <h5>有固定值（受控）</h5>
            <Radio.Group value={selected}>
                <Radio value="1">选项一</Radio>
                <Radio value="2">选项二</Radio>
                <Radio value="3">选项三</Radio>
                <Radio value="4">选项四</Radio>
            </Radio.Group>

            <h5>通用方法</h5>
            <Radio.Group value={selected} onChange={handleChange}>
                <Radio value="1">选项一</Radio>
                <Radio value="2">选项二</Radio>
                <Radio value="3">选项三</Radio>
                <Radio value="4">选项四</Radio>
            </Radio.Group>
        </div>
    );
}
```

## 单选框（按钮型）

```jsx live fff
<div className="demo-row">
    <Radio.Button checked type="radio-button">默认选中</Radio.Button>
    <Radio.Button type="radio-button">无默认值</Radio.Button>
</div>
```

## 单选框组（按钮型）

```jsx live fff
<Radio.Group>
    <Radio.Button value="1">选项一</Radio.Button>
    <Radio.Button value="2">选项二</Radio.Button>
    <Radio.Button value="3">选项三</Radio.Button>
    <Radio.Button value="4">选项四</Radio.Button>
</Radio.Group>
```

## 限制 label 长度

按钮形式的单选框可以对 `label` 进行限制，当长度超过指定长度时，会自动截断。

```jsx live fff
<div className="demo-column">
    <Radio limit={6}>123456789</Radio.Button>
    <Radio.Button limit={6}>123456789</Radio.Button>

    <Radio.Group>
        <Radio value="1">123456789</Radio>
        <Radio value="2">123456789</Radio>
        <Radio value="3">123456789</Radio>
        <Radio value="4">123456789</Radio>
    </Radio.Group>

    <Radio.Group>
        <Radio.Button value="1">123456789</Radio.Button>
        <Radio.Button value="2">123456789</Radio.Button>
        <Radio.Button value="3">123456789</Radio.Button>
        <Radio.Button value="4">123456789</Radio.Button>
    </Radio.Group>
</div>
```

## 传入 options 的用法

```jsx live fff
function App() {
    const options = [{
        value: 'js',
        label: 'JS'
    }, {
        value: 'css',
        label: 'CSS'
    }, {
        value: 'python',
        label: 'PYTHON'
    }, {
        value: 'java',
        label: 'JAVA'
    }, {
        value: 'go',
        label: 'GO'
    }, {
        value: 'ts',
        label: 'TS'
    }, {
        value: 'html',
        label: 'HTML'
    }];

    const [selected, setSelected] = useState("ts");

    const handleChange = (e) => {
        console.log(e);
        setSelected(e.target.value);
    };

    return (
        <div className="demo-column">
            <h5>一般 options 用法</h5>
            <Radio.Group
                value={selected}
                options={options}
                onChange={handleChange}
            />

            <h5>options + 按钮形式</h5>
            <Radio.Group
                optionType="button"
                value={selected}
                options={options}
                onChange={handleChange}
            />
        </div>
    );
}
```

## 禁用状态

```jsx live fff
function App() {
    const options = [{
        value: 'js',
        label: 'JS',
    }, {
        value: 'css',
        label: 'CSS',
    }, {
        value: 'python',
        label: 'PYTHON',
    }, {
        value: 'java',
        label: 'JAVA',
        disabled: true,
    }, {
        value: 'go',
        label: 'GO',
    }, {
        value: 'ts',
        label: 'TS',
    }, {
        value: 'html',
        label: 'HTML',
    }];

    const [selected, setSelected] = useState('ts');

    const handleChange = (e) => {
        console.log(e);
        setSelected(e.target.value);
    };

    return (
        <div className="demo-column">
            <h5>单项禁用</h5>
            <Radio disabled>未选中</Radio>
            <Radio.Button disabled>
                未选中
            </Radio.Button>

            <h5>单选框组局部禁用</h5>
            <Radio.Group value={selected} onChange={handleChange}>
                <Radio value="1">选项一</Radio>
                <Radio value="2" disabled>选项二</Radio>
                <Radio value="3">选项三</Radio>
            </Radio.Group>
            <Radio.Group value={selected} onChange={handleChange}>
                <Radio.Button value="1">选项一</Radio.Button>
                <Radio.Button value="2">选项二</Radio.Button>
                <Radio.Button value="3" disabled>选项三</Radio.Button>
            </Radio.Group>
            <Radio.Group
                value={selected}
                options={options}
                onChange={handleChange}
            />

            <h5>单选框组全局禁用</h5>
            <Radio.Group disabled value={selected} onChange={handleChange}>
                <Radio value="1">选项一</Radio>
                <Radio value="2">选项二</Radio>
                <Radio value="3">选项三</Radio>
            </Radio.Group>
            <Radio.Group disabled value={selected} onChange={handleChange}>
                <Radio.Button value="1">选项一</Radio.Button>
                <Radio.Button value="2">选项二</Radio.Button>
                <Radio.Button value="3">选项三</Radio.Button>
            </Radio.Group>
            <Radio.Group
                disabled
                optionType="button"
                value={selected}
                options={options}
                onChange={handleChange}
            />
        </div>
    );
}
```

## API

### Radio/RadioButton

| 参数      | 说明                              | 类型      | 默认值  |
| --------- | --------------------------------- | --------- | ------- |
| checked   | 是否选中                          | `boolean` | `false` |
| disabled  | 是否禁用                          | `boolean` | `false` |
| value     | 根据 value 进行比较，判断是否选中 | `any`     | `--`    |
| className | 自定义类名                        | `string`  | `--`    |

### RadioButton

| 参数  | 说明                      | 类型                  | 默认值 |
| ----- | ------------------------- | --------------------- | ------ |
| limit | 控制 label 超长时如何展示 | `number` \| `boolean` | `8`    |

:::caution 注意

- 值为 `false` 或 `0` 时，自适应宽度；
- 非零正整数，超过指定字符后省略展示；
- 其他情况，超过 `8` 字符省略展示。

:::

### RadioGroup

多用于在一组选项中进行单项选择且选项互斥，一般需要提交的选择型控件

| 参数       | 说明                        | 类型                       | 默认值    |
| ---------- | --------------------------- | -------------------------- | --------- |
| options    | 以配置形式设置子元素        | `string[]` \| `Array<...>` | `--`      |
| optionType | 用于设置 Radio options 类型 | `default`\|`button`        | `default` |
| value      | 用于设置当前选中的值        | `any`                      | `--`      |
| disabled   | 禁选所有子单选器            | `boolean`                  | `false`   |
| className  | 自定义类名                  | `string`                   | `--`      |
| enhance    | 增强样式，按钮样式时生效    | `boolean`                  | `false`   |
| onChange   | 选项变化时的回调函数        | `function(e: Event))`      | `--`      |
