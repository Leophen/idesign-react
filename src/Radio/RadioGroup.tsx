import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import './index.scss';
import { RadioContextValue, RadioGroupProps, RadioProps } from './type';

export const RadioContext = React.createContext<RadioContextValue>(null as any);

// 单选框组
const RadioGroup: React.FC<RadioGroupProps> = (props) => {
  const {
    children = '',
    className,
    style,
    currentValue,
    type = 'radio',
    size,
    disabled,
    onChange,
  } = props;

  const [groupValue, setGroupValue] = useState(currentValue);

  useEffect(() => {
    setGroupValue(currentValue);
  }, [currentValue]);

  // 注入每一项单选框的 context
  const context: RadioContextValue = {
    // 将单选框组的 props 注入单项单选框的方法
    inject: (singleRadioProps: RadioProps) => {
      // 拿到单项单选框的 value，方便与单选框组的 groupValue 做比较
      const singleRadioValue = singleRadioProps.value;
      return {
        ...singleRadioProps,
        type,
        size,
        disabled,
        checked: groupValue === singleRadioValue,
        onChange(checked, e) {
          e.persist();
          // 触发单项单选框传入的 onChange
          if (typeof singleRadioProps.onChange === 'function') {
            singleRadioProps.onChange(checked, e);
          }
          // 触发单选框组传入的 onChange
          singleRadioValue && onChange?.(singleRadioValue, e);
          // 设置单选框组的 currentValue
          setGroupValue(singleRadioValue);
        },
      };
    },
  };

  return (
    <RadioContext.Provider value={context}>
      <div className={classNames('i-radio-group', className)} style={{ ...style }}>
        {children}
      </div>
    </RadioContext.Provider>
  );
};

export default RadioGroup;
