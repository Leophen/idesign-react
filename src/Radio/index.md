---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 2
---

# Radio 单选框

单选框代表从一组互斥的选项中仅选择一个选项。

## 单选框

最简单的单选框形式。

```tsx
import React, { useState } from 'react';
import { Radio } from 'idesign-react';

const App = () => {
  const [checked, setChecked] = useState(false);

  const handleChange = (val, e) => {
    console.log(val, e);
    setChecked(val);
  };

  return (
    <div className="idesign-demo-block-row">
      <Radio checked={true}>已选中</Radio>
      <Radio checked={checked} onChange={handleChange}>
        {checked ? '已选中' : '未选中'}
      </Radio>
    </div>
  );
};

export default App;
```

## 单选框组

使用 `Radio.Group` 可以包裹 `Radio` 形成一组单选框组。

```tsx
import React, { useState } from 'react';
import { Radio } from 'idesign-react';

const App = () => {
  const [currentValue, setCurrentValue] = useState('2');

  const handleChange = (val, e) => {
    console.log(val, e);
    setCurrentValue(val);
  };

  return (
    <Radio.Group currentValue={currentValue} onChange={handleChange}>
      <Radio value="1">选项一</Radio>
      <Radio value="2">选项二</Radio>
      <Radio value="3">选项三</Radio>
      <Radio value="4">选项四</Radio>
    </Radio.Group>
  );
};

export default App;
```

## 单选框（按钮型）

`Radio` 提供 `type` 属性切换为按钮形式的单选框。

```tsx
import React, { useState } from 'react';
import { Radio } from 'idesign-react';

const App = () => {
  const [checked, setChecked] = useState(false);

  const handleChange = (val, e) => {
    console.log(val, e);
    setChecked(val);
  };

  return (
    <div className="idesign-demo-block-row">
      <Radio checked={true} type="radio-button">
        已选中
      </Radio>
      <Radio checked={checked} onChange={handleChange} type="radio-button">
        {checked ? '已选中' : '未选中'}
      </Radio>
    </div>
  );
};

export default App;
```

## 单选框组（按钮型）

`Radio.Group` 提供 `type` 属性全局修改每一项为按钮型单选框。

```tsx
import React, { useState } from 'react';
import { Radio } from 'idesign-react';

const App = () => {
  const [currentValue, setCurrentValue] = useState('2');

  const handleChange = (val, e) => {
    console.log(val, e);
    setCurrentValue(val);
  };

  return (
    <Radio.Group type="radio-button" currentValue={currentValue} onChange={handleChange}>
      <Radio value="1">选项一</Radio>
      <Radio value="2">选项二</Radio>
      <Radio value="3">选项三</Radio>
      <Radio value="4">选项四</Radio>
    </Radio.Group>
  );
};

export default App;
```

## 不同尺寸（按钮型）

`Radio` 按钮形式的单选框提供 `size` 属性控制尺寸，`Radio.Group` 也有 `size` 属性可以全局控制每一项按钮单选框的尺寸。

```tsx
import React, { useState } from 'react';
import { Radio } from 'idesign-react';

const App = () => {
  const [currentValue, setCurrentValue] = useState('2');

  const handleChange = (val) => {
    setCurrentValue(val);
  };

  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Radio type="radio-button" size="small">
          未选中
        </Radio>
        <Radio type="radio-button" size="medium" checked={true}>
          未选中
        </Radio>
        <Radio type="radio-button" size="large">
          未选中
        </Radio>
      </div>

      <div className="idesign-demo-block-row">
        <Radio.Group
          type="radio-button"
          size="small"
          currentValue={currentValue}
          onChange={handleChange}
        >
          <Radio value="1">选项一</Radio>
          <Radio value="2">选项二</Radio>
          <Radio value="3">选项三</Radio>
        </Radio.Group>
      </div>

      <div className="idesign-demo-block-row">
        <Radio.Group
          type="radio-button"
          size="medium"
          currentValue={currentValue}
          onChange={handleChange}
        >
          <Radio value="1">选项一</Radio>
          <Radio value="2">选项二</Radio>
          <Radio value="3">选项三</Radio>
        </Radio.Group>
      </div>

      <div className="idesign-demo-block-row">
        <Radio.Group
          type="radio-button"
          size="large"
          currentValue={currentValue}
          onChange={handleChange}
        >
          <Radio value="1">选项一</Radio>
          <Radio value="2">选项二</Radio>
          <Radio value="3">选项三</Radio>
        </Radio.Group>
      </div>
    </div>
  );
};

export default App;
```

## 禁用状态

`Radio` 提供 `disabled` 属性控制单选框为禁用状态，`Radio.Group` 可以全局控制每一项为禁用状态。

```tsx
import React, { useState } from 'react';
import { Radio } from 'idesign-react';

const App = () => {
  const [currentValue, setCurrentValue] = useState('2');

  const handleChange = (val) => {
    setCurrentValue(val);
  };

  return (
    <div className="idesign-demo-block-column">
      <h3>单项禁用</h3>

      <div className="idesign-demo-block-row">
        <Radio disabled>未选中</Radio>
        <Radio type="radio-button" disabled>
          未选中
        </Radio>
      </div>

      <h3>单选框组局部禁用</h3>

      <div className="idesign-demo-block-row">
        <Radio.Group currentValue={currentValue} onChange={handleChange}>
          <Radio value="1">选项一</Radio>
          <Radio value="2" disabled>
            选项二
          </Radio>
          <Radio value="3">选项三</Radio>
        </Radio.Group>
      </div>
      <div className="idesign-demo-block-row">
        <Radio.Group type="radio-button" currentValue={currentValue} onChange={handleChange}>
          <Radio value="1">选项一</Radio>
          <Radio value="2">选项二</Radio>
          <Radio value="3" disabled>
            选项三
          </Radio>
        </Radio.Group>
      </div>

      <h3>单选框组全局禁用</h3>

      <div className="idesign-demo-block-row">
        <Radio.Group disabled currentValue={currentValue} onChange={handleChange}>
          <Radio value="1">选项一</Radio>
          <Radio value="2">选项二</Radio>
          <Radio value="3">选项三</Radio>
        </Radio.Group>
      </div>

      <div className="idesign-demo-block-row">
        <Radio.Group
          disabled
          type="radio-button"
          currentValue={currentValue}
          onChange={handleChange}
        >
          <Radio value="1">选项一</Radio>
          <Radio value="2">选项二</Radio>
          <Radio value="3">选项三</Radio>
        </Radio.Group>
      </div>
    </div>
  );
};

export default App;
```

## Radio API

<API hideTitle />

## RadioGroup API

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| children | 按钮内容 | `ReactNode` | `--` |
| className | 类名 | `string` | `--` |
| style | 自定义样式 | `CSSProperties` | `--` |
| currentValue | 单选框组选中的值 | `string〡number` | `--` |
| type | 单选框组全局类型 | `"radio"〡"radio-button"` | `"radio"` |
| size | 按钮单选框全局尺寸 | `"small"〡"medium"〡"large"` | `"medium"` |
| disabled | 单选框组是否全局禁用 | `boolean` | `false` |
| onChange | 选中某一项时触发 | `(value: boolean〡number, context: React.ChangeEvent<HTMLInputElement>) => void` | `--` |
