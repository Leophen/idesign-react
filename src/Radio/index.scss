@keyframes checked {
  0% {
    transform: scale(0);
  }
  50% {
    transform: scale(1.2);
  }
  100% {
    transform: scale(1);
  }
}

.i-radio {
  position: relative;
  cursor: pointer;
  user-select: none;
  &.i-radio-is-checked {
    .i-radio__input {
      border-color: var(--i-primary);
      &::after {
        transform: scale(1);
        animation: checked 0.2s cubic-bezier(0, 0, 0.15, 1);
      }
    }
  }
  &:hover {
    .i-radio__input {
      border-color: var(--i-primary);
    }
  }
  &:active {
    &.i-radio-is-checked {
      .i-radio__input {
        &::after {
          transform: scale(1.1);
        }
      }
    }
  }
  &.i-radio-is-disabled {
    cursor: not-allowed;
    .i-radio__input {
      background-color: var(--i-bg-disabled);
      &::after {
        background-color: var(--i-border-hover);
      }
    }
    &.i-radio-is-checked {
      .i-radio__input {
        border-color: var(--i-border);
      }
      &:hover {
        .i-radio__input {
          border-color: var(--i-border);
        }
      }
    }
    &:hover {
      .i-radio__input {
        border-color: var(--i-border);
      }
    }
  }
}

.i-radio__input {
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  width: 16px;
  height: 16px;
  vertical-align: middle;
  background-color: var(--i-bg);
  border: 1px solid var(--i-border);
  border-radius: 50%;
  transition: border 0.2s cubic-bezier(0, 0, 0.15, 1);
  &::after {
    position: absolute;
    top: 50%;
    left: 50%;
    box-sizing: border-box;
    width: 8px;
    height: 8px;
    margin-top: -4px;
    margin-left: -4px;
    background-color: var(--i-primary);
    border-radius: 50%;
    transform: scale(0);
    transition: all 0.2s cubic-bezier(0, 0, 0.15, 1);
    content: '';
  }
}

.i-radio__label {
  display: inline-block;
  margin-left: 8px;
  color: var(--i-font);
  font-size: 14px;
  white-space: nowrap;
  vertical-align: middle;
}

.i-radio-button {
  position: relative;
  padding: 4px 16px;
  color: var(--i-font);
  font-size: 14px;
  white-space: nowrap;
  border: 1px solid var(--i-border);
  border-radius: 4px;
  cursor: pointer;
  transition: color 0.2s cubic-bezier(0, 0, 0.15, 1);
  user-select: none;
  &.i-radio-button-is-checked {
    color: var(--i-primary);
    border-color: var(--i-primary);
    &:active {
      color: var(--i-primary-active);
      border-color: var(--i-primary-active);
    }
  }
  &.i-radio-button--size-small {
    padding: 2px 8px;
    font-size: 12px;
  }
  &.i-radio-button--size-large {
    padding: 8px 24px;
    font-size: 16px;
  }
  &:hover {
    color: var(--i-primary-hover);
  }
  &:active {
    color: var(--i-primary);
  }
  &.i-radio-button-is-disabled {
    color: var(--i-font-disabled);
    background-color: var(--i-bg-disabled);
    border-color: var(--i-border) !important;
    cursor: not-allowed;
    &.i-radio-button-is-checked {
      color: var(--i-font-5);
      background-color: var(--i-bg-active);
      border-color: var(--i-border);
      &:hover {
        color: var(--i-font-5);
      }
    }
    &:hover {
      color: var(--i-font-disabled);
    }
    &:active {
      color: var(--i-font-disabled);
    }
  }
}

.i-radio__former,
.i-radio-button__former {
  position: absolute;
  width: 0;
  height: 0;
  opacity: 0;
}

.i-radio-group {
  display: inline-flex;
  align-items: center;
  .i-radio {
    margin-right: 24px;
  }
  .i-radio-button {
    border-right: 0;
    border-radius: 0;
    &:first-of-type {
      border-radius: 4px 0 0 4px;
    }
    &:last-of-type {
      border-right: 1px solid var(--i-border);
      border-radius: 0 4px 4px 0;
    }
    &.i-radio-button-is-checked {
      border-top-color: var(--i-primary);
      border-right: 1px solid;
      border-right-color: var(--i-primary);
      border-bottom-color: var(--i-primary);
      border-left-color: var(--i-primary);
      transition: border-color 0.2s cubic-bezier(0, 0, 0.15, 1);
      & + .i-radio-button {
        border-left: 0;
      }
    }
  }
}
