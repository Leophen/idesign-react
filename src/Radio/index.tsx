import React, { useState, useContext, useEffect } from 'react';
import classNames from 'classnames';
import './index.scss';
import RadioGroup from './RadioGroup';
import { RadioContextValue, RadioProps } from './type';

export const RadioContext = React.createContext<RadioContextValue>(null as any);

// 单一单选框
const Radio: React.FC<RadioProps> & { Group: React.ElementType } = (props) => {
  // 存在单选框组时从 Context 注入 checked 覆盖原 checked
  const context = useContext(RadioContext);
  const newProps = context ? context.inject(props) : props;

  const {
    children = '',
    className,
    style,
    checked = false,
    value,
    onChange,
    ...restProps
  } = newProps;

  const type = newProps.type || props.type || 'radio';
  const size = newProps.size || props.size;
  const disabled = newProps.disabled || props.disabled || false;

  const [innerChecked, setInnerChecked] = useState(checked);

  useEffect(() => {
    setInnerChecked(checked);
  }, [checked]);

  const handleCheck = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (disabled) return;
    const checkedValue = e.currentTarget.checked;
    // 单选框组时使用 context 提供的修改 checked 方法同步整组的 currentValue
    if (typeof newProps.onChange === 'function') {
      newProps.onChange(checkedValue, e);
    }
  };

  return (
    <label
      className={classNames(
        `i-${type}`,
        innerChecked && `i-${type}-is-checked`,
        disabled && `i-${type}-is-disabled`,
        size && `i-${type}--size-${size}`,
        className,
      )}
      style={{ ...style }}
      {...restProps}
    >
      <input
        readOnly
        type="radio"
        className={`i-${type}__former`}
        checked={innerChecked}
        disabled={disabled}
        value={value}
        onClick={(e) => e.stopPropagation()}
        onChange={(e) => handleCheck(e)}
      />
      <span className={`i-${type}__input`} />
      <span className={`i-${type}__label`}>{children}</span>
    </label>
  );
};

Radio.Group = RadioGroup;

Radio.displayName = 'Radio';
RadioGroup.displayName = 'RadioGroup';

export default Radio;
