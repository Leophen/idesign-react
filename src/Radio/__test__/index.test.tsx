import { render, fireEvent } from '@testing-library/react';
import Radio from '../index';

describe('Radio 组件测试', () => {
  const renderRadio = (box) => {
    const { container } = render(box);
    return container.firstChild
  }

  it('create', () => {
    const { asFragment } = render(<Radio />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('checked & children', () => {
    const { container, queryByText } = render(<Radio checked={true}>广州</Radio>);
    expect(container.firstChild).toHaveClass('i-radio', 'i-radio-is-checked');
    expect(queryByText('广州')).toBeInTheDocument();
  });

  it('defaultChecked', () => {
    expect(renderRadio(<Radio defaultChecked={true} />)).toHaveClass('i-radio', 'i-radio-is-checked');
  });

  it('disabled', () => {
    const fn = jest.fn();
    const radio = renderRadio(<Radio disabled={true} onChange={fn} />)
    expect(radio).toHaveClass('i-radio-is-disabled');
    fireEvent.click(radio);
    expect(fn).toBeCalledTimes(0);
  });

  it('type', () => {
    expect(renderRadio(<Radio type="radio-button" />)).toHaveClass('i-radio-button');
  });

  it('size', () => {
    expect(renderRadio(<Radio type="radio-button" size="small" />)).toHaveClass('i-radio-button--size-small');
  });
});

describe('RadioGroup 组件测试', () => {
  it('selected', () => {
    const { getByTestId } = render(
      <Radio.Group selected='sz'>
        <Radio value="gz">广州</Radio>
        <Radio value="sz" data-testid='sz-item'>深圳</Radio>
      </Radio.Group>,
    );
    expect(getByTestId('sz-item')).toHaveClass('i-radio-is-checked');
  });

  it('defaultSelected', () => {
    const { getByTestId } = render(
      <Radio.Group defaultSelected='gz'>
        <Radio value="gz" data-testid='gz-item'>广州</Radio>
        <Radio value="sz">深圳</Radio>
      </Radio.Group>,
    );
    expect(getByTestId('gz-item')).toHaveClass('i-radio-is-checked');
  });

  it('disabled', () => {
    const fn = jest.fn();
    const { getByTestId } = render(
      <Radio.Group disabled>
        <Radio value="gz" data-testid='gz-item'>广州</Radio>
        <Radio value="sz">深圳</Radio>
      </Radio.Group>,
    );
    expect(getByTestId('gz-item')).toHaveClass('i-radio-is-disabled');
    fireEvent.click(getByTestId('gz-item'));
    expect(fn).toBeCalledTimes(0);
  });

  it('size', () => {
    const { getByTestId } = render(
      <Radio.Group type="radio-button" size="small">
        <Radio value="gz" data-testid='gz-item'>广州</Radio>
        <Radio value="sz">深圳</Radio>
      </Radio.Group>,
    );
    expect(getByTestId('gz-item')).toHaveClass('i-radio-button--size-small');
  });
});

