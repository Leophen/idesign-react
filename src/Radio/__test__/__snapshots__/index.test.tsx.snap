// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio 组件测试 create 1`] = `
<DocumentFragment>
  <label
    class="i-radio"
  >
    <input
      class="i-radio__former"
      readonly=""
      type="radio"
      value=""
    />
    <span
      class="i-radio__input"
    />
    <span
      class="i-radio__label"
    />
  </label>
</DocumentFragment>
`;
